### PR TITLE
bug: Fix `TypeExpression::allowsNull()` with nullable

### PR DESCRIPTION
**Pull request by mvorisek**
No test failed after the fix... But `allowsNull` method name implies it should work with nullable phpdoc?

### CODE REVIEW

**Comment by mvorisek** [comment]
@Wirone do you think this is welcomed and do you have an idea of some functional test to add where `TypeExpression::allowsNull` result is used?

File: tests/DocBlock/TypeExpressionTest.php
Comment (Wirone) [inline tests/DocBlock/TypeExpressionTest.php:423]: From [RFC](https://wiki.php.net/rfc/union_types_v2#nullable_union_types):

> Union types and the `?T` nullable type notation cannot be mixed

It ends up with `Parse error: syntax error, unexpected token "?"`.

**Review by Wirone — CHANGES_REQUESTED** [review]
I believe `?something` should be supported by `allowsNull()` and covered by tests, so 👍 in general, but second test case with union type must be removed because it's not supported in PHP.

PS. `#7000` for such change 😅.

File: tests/DocBlock/TypeExpressionTest.php
Comment (mvorisek) [reply tests/DocBlock/TypeExpressionTest.php]: Yeah, `#7000`!

Test fixed, makes sense to no add someting not natively supported.

Comment (mvorisek) [reply tests/DocBlock/TypeExpressionTest.php]: To my question about functional case, do you think we can add some functional/fixer test that will fail /wo this fix?

Comment (Wirone) [reply tests/DocBlock/TypeExpressionTest.php]: `TypeExpression::allowsNull()` is used only in `AbstractPhpdocToTypeDeclarationFixer::getCommonTypeFromAnnotation()`, so probably test cases should be added to its child classes: `PhpdocToReturnTypeFixer`, `PhpdocToParamTypeFixer` and `PhpdocToPropertyTypeFixer`. Currently it fails, because in `AbstractPhpdocToTypeDeclarationFixer::getCommonTypeFromAnnotation()` there's check `Preg::match(self::CLASS_REGEX, $commonType)` which causes that `null` is returned and fix is skipped.

image

I rather don't see `?Foo` format usage in phpDoc (only union), but [it's valid too](https://phpstan.org/r/ed03e113-3624-4b3d-998d-225b48367d17), so it should be covered. Would like to fix `AbstractPhpdocToTypeDeclarationFixer` and provide test cases for those 3 mentioned fixers 🙂?

Comment (mvorisek) [reply tests/DocBlock/TypeExpressionTest.php]: Please push the tests to my PR branch.

File: tests/DocBlock/TypeExpressionTest.php
Comment (Wirone) [reply tests/DocBlock/TypeExpressionTest.php:369]: Something with e.g. `['foo']`, but I am not sure if we really need it? LGTM to me as-is.

**Review by Wirone — APPROVED** [review]
@kubawerlos please take a quick look 🙏.

File: tests/DocBlock/TypeExpressionTest.php
Comment (mvorisek) [reply tests/DocBlock/TypeExpressionTest.php:369]: I mean some E2E test which can be added to some fixer testcases.

**Comment by Wirone** [comment]
@mvorisek FYI: @kubawerlos won't be able to look at it in the next 2 weeks, so if any other maintainer doesn't look at it, we'll just wait so please be patient 🥲.

**Review by kubawerlos — APPROVED** [review]

**Comment by Wirone** [comment]
@mvorisek do you have Twitter/Mastodon account? 🙂

**Comment by mvorisek** [comment]
@Wirone I am on Discord `mvorisek` (`mvorisek#3493`)

**Comment by Wirone** [comment]
> @Wirone I am on Discord `mvorisek` (`mvorisek#3493`) or on LinkedIn as Michael Voříšek

image

image

@mvorisek unfortunately both communication channels are unavailable for me. Can you send friend request to me (Discord: `wirone`, [LinkedIn](https://www.linkedin.com/in/grzegorz-korba/)) 🙂?

**Comment by mvorisek** [comment]
@Wirone, sent for both channels ❤, use Discord for anything (but I am there only time by time) or LinkedIn for more business or urgent communication.

**Pull request merged**
